### PR TITLE
support pulling multiple images sequentially in a single podman pull 

### LINKED ISSUE #10726: Podman should support pulling multiple images in a single command
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**
Currently podman pull command accepts only one argument and it would be really nice if it could support pulling images in a single command.

```
[mehul@fedora bin]$ ./podman pull alpine busybox
Error: accepts 1 arg(s), received 2
```
If someone wants to pull multiple images and if they are not familiar with bash scripting then they might have to do it manually by calling the command over and over. Instead it would be nice to have something like:
```
podman pull alpine busybox fedora ubuntu
```

**Output of `podman version`:**

```
[mehul@fedora bin]$ ./podman --version
podman version 3.3.0-dev
```
I think this could be a useful feature, but I am not completely sure about it's use case except I felt that it was redundant to type the command again when I was working on something.  I can try to work on this if this is agreed upon! 

Thank you

### PR DESCRIPTION
**Pull request by infiniteregrets**
Closes #10726 
Adds support for pulling multiple images sequentially. 

Signed-off-by: Mehul Arora <aroram18@mcmaster.ca>

### CODE REVIEW

**Comment by infiniteregrets** [comment]
@rhatdan PTAL

File: cmd/podman/images/pull.go
Comment (mtrmac) [inline cmd/podman/images/pull.go:178]: This needs finishing.

**Review by mtrmac — COMMENTED** [review]
> If an image exists, it is skipped. Not sure why this wasn't present before but podman tries to pull an existing image:

If the `:latest` tag on a repo changes, why _wouldn’t_ `podman pull` verify that it has the updated version?

File: cmd/podman/images/pull.go
Comment (infiniteregrets) [reply cmd/podman/images/pull.go:178]: > If the :latest tag on a repo changes, why wouldn’t podman pull verify that it has the updated version?

My bad, I didn't realize. Will fix that

Comment (infiniteregrets) [reply cmd/podman/images/pull.go:178]: > This needs finishing.

I forgot to ask, I didn't understand what the portion of code I commented is for?

File: cmd/podman/images/pull.go
Comment (vrothberg) [inline cmd/podman/images/pull.go:36]: Can you use `Args: cobra.MinimumNArgs(1)` instead?  This will do the same but be consistent with the error messages of other commands.

Comment (infiniteregrets) [reply cmd/podman/images/pull.go]: Oh right, thanks!

File: cmd/podman/images/pull.go
Comment (jwhonce) [reply cmd/podman/images/pull.go:165]: @infiniteregrets For each pull, you need to print out the image ids that were pulled.

**Review by jwhonce — CHANGES_REQUESTED** [review]

File: cmd/podman/images/pull.go
Comment (infiniteregrets) [reply cmd/podman/images/pull.go]: @jwhonce I was a bit confused because, "Pulled Images:" was never getting printed before only the signature was. I have made the changes, please let me know if this is okay now?

File: cmd/podman/images/pull.go
Comment (vrothberg) [inline cmd/podman/images/pull.go:164]: I assume the newline was added to have a blank line between each image pull but it will break `-q`.  I suggest to just remove the line.

File: cmd/podman/images/pull.go
Comment (vrothberg) [inline cmd/podman/images/pull.go:163]: `pullReport.Images` can have more than one image (see `podman pull -a`) so we should do as follows:
```go
var pulledImages []string
for i, arg := .... {
   ....
   pulledImages = append(pulledImages, pullReport.Images...)
}
```
This will append all pulled images.

```

File: test/e2e/pull_test.go
Comment (vrothberg) [inline test/e2e/pull_test.go:41]: We should also check the output string and make sure that the images are correctly reported as pulled.

Comment (vrothberg) [reply test/e2e/pull_test.go:41]: Look for `GrepString(...)` in the other tests.

File: cmd/podman/images/pull.go
Comment (infiniteregrets) [reply cmd/podman/images/pull.go:164]: Sure Ill remove it, but can we not do something like:
```go
if !pullOptions.Quiet {
    fmt.Print("\n")
}
```

Comment (vrothberg) [reply cmd/podman/images/pull.go:164]: The output of `pull --all foo` doesn't have these blank lines, so it would not be consistent.  But I do not feel strongly.  If you like the blank lines, feel free to add them.

File: cmd/podman/images/pull.go
Comment (rhatdan) [inline cmd/podman/images/pull.go:165]: Is this correct?  If I pull a bunch of images with quiet, doesn't it print the ImageID for each image I pull? Why doesn't that need a \n for more then one?

Comment (rhatdan) [reply cmd/podman/images/pull.go:165]: I would expect the output to be
```
$ podman pull ubi8-micro fedora
Trying to pull registry.access.redhat.com/ubi8-micro:latest...
Getting image source signatures
Copying blob c07a48b26be2 skipped: already exists  
Copying blob 4f4fb700ef54 [--------------------------------------] 0.0b / 0.0b
Copying config f390b26f6a done  
Writing manifest to image destination
Storing signatures
f390b26f6a00561dbacfc5991a39a6845ff7968f8d18a319b68cc9c0f97dea73
Trying to pull registry.fedoraproject.org/fedora:latest...
Getting image source signatures
Copying blob ca594071dfff skipped: already exists  
Copying config abec9a7a7d done  
Writing manifest to image destination
Storing signatures
abec9a7a7dc640768e6dc51b6a5728e470411615c62e9ff46215206bde816772
```
Rather then

```
$ podman pull ubi8-micro fedora
Trying to pull registry.access.redhat.com/ubi8-micro:latest...
Getting image source signatures
Copying blob c07a48b26be2 skipped: already exists  
Copying blob 4f4fb700ef54 [--------------------------------------] 0.0b / 0.0b
Copying config f390b26f6a done  
Writing manifest to image destination
Storing signatures
Trying to pull registry.fedoraproject.org/fedora:latest...
Getting image source signatures
Copying blob ca594071dfff skipped: already exists  
Copying config abec9a7a7d done  
Writing manifest to image destination
Storing signatures
Pulled Images:
f390b26f6a00561dbacfc5991a39a6845ff7968f8d18a319b68cc9c0f97dea73
abec9a7a7dc640768e6dc51b6a5728e470411615c62e9ff46215206bde816772
```

File: cmd/podman/images/pull.go
Comment (rhatdan) [inline cmd/podman/images/pull.go:161]: When we try to do multiple operations at once, we complete the tasks, and print the errors.  Look how podman stop works.
If I say

podman pull alpine bogus fedora ubi

I would want alpine fedora and ubi to be pulled, and then the command fail with bogus does not exist.

**Review by rhatdan — APPROVED** [review]

File: cmd/podman/images/pull.go
Comment (mtrmac) [reply cmd/podman/images/pull.go:165]: I don’t feel strongly about this, but the second one seems _much_ easier to copy&paste. (Of course users should not copy&paste non-API command output, but then why are the IDs printed at all?)

Comment (rhatdan) [reply cmd/podman/images/pull.go:165]: Because docker printed them, and they make sense when you do it with a -q option on one image.

File: cmd/podman/images/pull.go
Comment (infiniteregrets) [reply cmd/podman/images/pull.go:161]: I think that is possible because the ContainerStop function accepts an array of strings and here it's just a single string. And passing an array makes it possible to get an accumulated response but in this case we would have to change the code for Pull in various places. 
But if you think that this is possible, then this PR is editable by maintainers. So please make the change. Thanks!

File: cmd/podman/images/pull.go
Comment (infiniteregrets) [reply cmd/podman/images/pull.go:165]:
```
❯ docker pull -q busybox
docker.io/library/busybox:latest
```
Docker doesnt print the IDs for -q flag. 
So let me know what the desired behavior is and I will make the changes accordingly

File: cmd/podman/images/pull.go
Comment (rhatdan) [reply cmd/podman/images/pull.go:161]: No you just need something like:
```
var errs utils.OutputErrors
for _, arg := range args {
		pullReport, err := registry.ImageEngine().Pull(registry.GetContext(), arg, pullOptions.ImagePullOptions)
		if err != nil {
                        errs = append(errs, err)
			continue
              }
    ...
    }
    return errs.PrintErrors()
}

File: cmd/podman/images/pull.go
Comment (rhatdan) [reply cmd/podman/images/pull.go:165]: I would prefer to print the image name on success, but I think we have set the precedent of printing the ID.  WDYT @vrothberg @mheon ?

Comment (vrothberg) [reply cmd/podman/images/pull.go:165]: I think we should do what @rhatdan suggested in https://github.com/containers/podman/pull/10788#discussion_r660486435.

File: cmd/podman/images/pull.go
Comment (infiniteregrets) [reply cmd/podman/images/pull.go]: I have updated the code, but why arent the checks running?

Comment (vrothberg) [reply cmd/podman/images/pull.go]: Yes, it looks like a `master` -> `main` issue. Hm ... maybe try closing and re-opening. Then repush?

Comment (infiniteregrets) [reply cmd/podman/images/pull.go]: @vrothberg I fixed that, but some tests are failing. Am I missing something?

Comment (vrothberg) [reply cmd/podman/images/pull.go]: Tests are currently running.  I'll wait for them to finish and will have a look later.

Comment (infiniteregrets) [reply cmd/podman/images/pull.go]: Looks like it is unable to resolve the tag I am using for the alpine image
[Error: Error initializing source docker://alpine:edge: Error reading manifest edge in quay.io/libpod/alpine: manifest unknown: manifest unknown](https://storage.googleapis.com/cirrus-ci-6707778565701632-fcae48/artifacts/containers/podman/4715965364568064/html/int-podman-fedora-34-rootless-host.log.html)
But this works just fine locally for me?

File: test/e2e/pull_test.go
Comment (vrothberg) [inline test/e2e/pull_test.go:39]: `alpine:edge` does not exist on our mirror on quay.  Please use `alpine`.  Then you can also remove the `"podman pull multiple images without tag"` test.

File: cmd/podman/images/pull.go
Comment (rhatdan) [inline cmd/podman/images/pull.go:169]: No need for this check, if errs.PrintErrors will return nil if there are no errors.
LGTM

**Comment by rhatdan** [comment]
Nice work @infiniteregrets.  Thanks
LGTM
@containers/podman-maintainers PTAL

**Review by vrothberg — COMMENTED** [review]
/lgtm
/hold

Thanks!

**Comment by mheon** [comment]
Test failures seem consistent

File: cmd/podman/images/pull.go
Comment (Luap99) [inline cmd/podman/images/pull.go:36]: You have to add this, this will fix the failing test.
Also please update the synopsis in the man page.
```suggestion
		Use:               "pull [options] IMAGE [IMAGE...]",
```

**Comment by infiniteregrets** [comment]
hi @Luap99 let me know if there is anything to be changed in docs

**Comment by mheon** [comment]
LGTM

**Review by vrothberg — APPROVED** [review]
/lgtm

Thank you!

**Pull request merged**
